File: languages/at_archer/at_archer.js
"use strict";

const main = arg => {
  const input = arg.trim().split("\n");

  const n = parseInt(input[0].split(" ")[0]);
  const m = parseInt(input[0].split(" ")[1]);
  const d = parseInt(input[0].split(" ")[2]);
  const r = input[1].split(" ").map(x=>parseInt(x));
  const s = input[2].split(" ").map(x=>parseInt(x));
  console.log(solve(n, m, d, r, s));
}

let solve = function(n, m, d, r, s) {
  const a = -Math.floor(n * d / 2);
  const tmpr = r.map(x => x + 1).slice(1);
  const tmpList = r.reverse().map(x => -x).slice(0, -1).concat(tmpr);
  const l = tmpList.slice(0, -1);
  const rList = tmpList.slice(1);
  const tmps = s.slice(1);
  const sList = s.reverse().concat(tmps);
  const imos = Array(d + 1);
  imos.fill(0);
  for(let i = 0; i < 2 * m - 1; i++) {
    if(l[i] - a >= n * d) {
      imos[0] -= n * sList[i];
      imos[d] += n * sList[i];
    } else if(l[i] - a >= 1) {
      imos[0] -= Math.floor((l[i] - a + d) / d) * sList[i];
      imos[(l[i] - a) % d] += sList[i];
      imos[d] += Math.floor((l[i] - a) / d) * sList[i];
    }
    if(rList[i] - a >= n * d) {
      imos[0] += n * sList[i];
      imos[d] -= n * sList[i];
    } else if(rList[i] - a >= 1) {
      imos[0] += Math.floor((rList[i] - a + d) / d) * sList[i];
      imos[(rList[i] - a) % d] -= sList[i];
      imos[d] -= Math.floor((rList[i] - a) / d) * sList[i];
    }
  }
  let val = 0;
  let score = 0;
  for(let i = 0; i < d; i++) {
    val += imos[i];
    score = Math.max(score, val);
  }
  return score;
}

main(require('fs').readFileSync('/dev/stdin', 'utf8'));

File: languages/at_archer/at_archer.kt
import java.io.PrintWriter

fun main(args : Array<String>) {
    val pw = PrintWriter(System.out)
    val (n, m, d) = readLine()!!.split(" ").map{ it.toLong() }
    val r = readLine()!!.split(" ").map{ it.toLong() }
    val s = readLine()!!.split(" ").map{ it.toLong() }
    pw.println(solve(n, m, d, r, s))
    pw.flush()
}

fun solve(n: Long, m: Long, d: Long, r: List<Long>, s: List<Long>): Long {
    val a = -(n * d / 2)
    val tmp_list = r.asReversed().subList(0, m.toInt()).map { -it }.plus(r.subList(1, m.toInt() + 1).map { it + 1 })
    val l = tmp_list.subList(0, 2 * m.toInt() - 1)
    val r_list = tmp_list.subList(1, 2 * m.toInt())
    val s_list = s.asReversed().plus(s.subList(1, m.toInt()))
    var imos = Array(d.toInt() + 1) { 0L }
    for(i in 0 until 2 * m.toInt() - 1) {
        if(l[i] - a >= n * d) {
            imos[0] -= n * s_list[i];
            imos[d.toInt()] += n * s_list[i];
        } else if(l[i] - a >= 1) {
            imos[0] -= (l[i] - a + d) / d * s_list[i];
            imos[((l[i] - a) % d).toInt()] += s_list[i];
            imos[d.toInt()] += (l[i] - a) / d * s_list[i];
        }
        if(r_list[i] - a >= n * d) {
            imos[0] += n * s_list[i];
            imos[d.toInt()] -= n * s_list[i];
        } else if(r_list[i] - a >= 1) {
            imos[0] += (r_list[i] - a + d) / d * s_list[i];
            imos[((r_list[i] - a) % d).toInt()] -= s_list[i];
            imos[d.toInt()] -= (r_list[i] - a) / d * s_list[i];
        }
    }
    var value = 0L
    var score = 0L
    for(i in 0 until d.toInt()) {
        value += imos[i]
        if(value > score) {
            score = value
        }
    }
    return score
}